chain/chain/src/apply_chunks_thread_pool.rs

```diff
@@ -107,13 +94,53 @@ impl AsyncComputationSpawner for ThreadPool {
     }
 }
 
+/// This struct tracks the current number of worker threads
+/// and updates relevant metrics accordingly.
+struct WorkerCounter {
+    name: &'static str,
+    limit: usize,
+    num_threads: AtomicUsize,
+}
+
+impl WorkerCounter {
+    fn new(name: &'static str, limit: usize) -> Arc<Self> {
+        Arc::new(Self { name, limit, num_threads: Default::default() })
+    }
+
+    fn dec(&self) {
+        self.num_threads.fetch_sub(1, Ordering::SeqCst);
+        THREAD_POOL_NUM_THREADS.with_label_values(&[self.name]).dec();
+    }
+
+    fn new_thread(self: &Arc<Self>) -> WorkerCounterGuard {
```

Comment:
I was kind of expecting two separate structs, one for the counter and one for the guard with RAII with Arc to the counter. Do you think it would be any cleaner? I'm not against this one, it just took me by surprise.

Reply:
What would the guard hold if not some kind of reference to the counter? Should I copy all the references inside instead?

Reply:
This is what I was thinking but again no biggie:
```
Counter { num_threads: usize }
Guard { counter: Arc<counter> } 
```

Reply:
Okay, I think I will stick to the existing design.